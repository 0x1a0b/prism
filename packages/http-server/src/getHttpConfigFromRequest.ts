import { PrismConfig, PrismConfigFactory } from '@stoplight/prism-core/types';
import { IHttpConfig, IHttpOperationConfig, IHttpRequest } from '@stoplight/prism-http/types';

async function getConfig<Config, Input>(
  input: Input,
  config: PrismConfig<Config, Input>,
  defaultConfig?: PrismConfig<Config, Input>
): Promise<Config> {
  if (typeof config === 'function') {
    // config factory function
    return await (config as PrismConfigFactory<Config, Input>)(input, defaultConfig);
  }
  return config as Config;
}

export const getHttpConfigFromRequest: PrismConfigFactory<IHttpConfig, IHttpRequest> = async (
  req: IHttpRequest,
  defaultConfig?: PrismConfig<IHttpConfig, IHttpRequest>
) => {
  const config: IHttpConfig = defaultConfig
    ? await getConfig<IHttpConfig, IHttpRequest>(req, defaultConfig)
    : { mock: true };
  const query = req.url.query;
  if (!query) {
    return config;
  }

  const { __code, __dynamic, __contentType, __example } = query;

  const httpOperationConfig: IHttpOperationConfig = {
    code: __code,
    dynamic: __dynamic && __dynamic.toLowerCase() === 'true' ? true : false,
    mediaType: __contentType,
    exampleKey: __example,
  };

  if (Object.keys(httpOperationConfig).length) {
    if (typeof config.mock === 'boolean') {
      config.mock = httpOperationConfig;
    } else {
      config.mock = Object.assign({}, config.mock, httpOperationConfig);
    }
    return config;
  }

  return config;
};
