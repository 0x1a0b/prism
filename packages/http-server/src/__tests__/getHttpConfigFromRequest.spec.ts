import { IHttpRequest } from '@stoplight/prism-http';
import { getHttpConfigFromRequest } from '../getHttpConfigFromRequest';

describe('getHttpConfigFromRequest()', () => {
  test('extracts code', () => {
    return expect(
      getHttpConfigFromRequest({
        method: 'get',
        url: { path: '/', query: { __code: '202' } },
      } as IHttpRequest)
    ).resolves.toMatchSnapshot();
  });
  test('extracts mediaType', () => {
    return expect(
      getHttpConfigFromRequest({
        method: 'get',
        url: { path: '/', query: { __contentType: 'application/json' } },
      } as IHttpRequest)
    ).resolves.toMatchSnapshot();
  });
  test('extracts example', () => {
    return expect(
      getHttpConfigFromRequest({
        method: 'get',
        url: { path: '/', query: { __example: 'bear' } },
      } as IHttpRequest)
    ).resolves.toMatchSnapshot();
  });
  test('extracts dynamic', () => {
    return expect(
      getHttpConfigFromRequest({
        method: 'get',
        url: { path: '/', query: { __dynamic: 'true' } },
      } as IHttpRequest)
    ).resolves.toMatchSnapshot();
  });
});
