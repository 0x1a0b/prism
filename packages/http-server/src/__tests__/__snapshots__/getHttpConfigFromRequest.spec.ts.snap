// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHttpConfigFromRequest() extracts code 1`] = `
Object {
  "code": "202",
}
`;

exports[`getHttpConfigFromRequest() extracts dynamic 1`] = `
Object {
  "dynamic": true,
}
`;

exports[`getHttpConfigFromRequest() extracts example 1`] = `
Object {
  "exampleKey": "bear",
}
`;

exports[`getHttpConfigFromRequest() extracts mediaType 1`] = `
Object {
  "mediaType": "application/json",
}
`;
